### PR TITLE
Append the shard top docs in such a way to prevent AOOBE

### PR DESCRIPTION
**Pull request by martijnvg**
This way a ArrayIndexOutOfBoundsException like is reported in #7926 is impossible to occur.

### CODE REVIEW

**Comment by jpountz** [comment]
LGTM.

**Comment by jpountz** [comment]
Given that we are seeing such weird bugs, maybe we should check invariants more aggressively, such as ensuring that the same index of the `AtomicArray` is not set twice or that set is never called with `null`? What do you think? (Could be another PR)

**Comment by martijnvg** [comment]
@jpountz I totally agree

File: src/main/java/org/elasticsearch/action/search/type/TransportSearchScrollScanAction.java [hunk 1]
```diff
@@ -159,6 +161,10 @@ void executePhase(final int shardIndex, DiscoveryNode node, final long searchId)
             searchService.sendExecuteScan(node, internalScrollSearchRequest(searchId, request), new SearchServiceListener<QueryFetchSearchResult>() {
                 @Override
                 public void onResult(QueryFetchSearchResult result) {
+                    Objects.requireNonNull(result, "QueryFetchSearchResult can't be null");
+                    if (queryFetchResults.get(shardIndex) != null) {
+                        throw new IllegalStateException("Shard index [" + shardIndex + "] has already been set");
+                    }
                     queryFetchResults.set(shardIndex, result);
```
Comment (jpountz) [inline src/main/java/org/elasticsearch/action/search/type/TransportSearchScrollScanAction.java]: I think we should try to do it in an atomic way?

```java
if (queryFetchResults.compareAndSet(shardIndex, null, result) == false) {
    throw new Exception();
}
```

**Comment by jpountz** [comment]
LGTM

**Pull request merged**